feat: Add MySQL support

akashgangil opened the pull request:
Introduces a new variable database_type to configure
the three tier web app to use either MySQL or PostgreSQL.

Resuses gcr.io/sic-container-repo/todo-api, due to this MySQL
instance authentication is configured via a databases user/password
instead of IAM service account like PG. 

**MySQL**
```
module "three_tier_app" {
  source                 = "GoogleCloudPlatform/three-tier-web-app/google"
  version                = "~> 0.1"
  project_id             = <dummy>
  database_type          = "mysql"
  region                 = "us-central1"
  zone                   = "us-central1-a"
}
```

**PostgreSQL**
```
module "three_tier_app" {
  source                 = "GoogleCloudPlatform/three-tier-web-app/google"
  version                = "~> 0.1"
  project_id             = <dummy>
  database_type          =  "postgres"
  region                 = "us-central1"
  zone                   = "us-central1-a"
}
```

NimJay commented on main.tf:
```diff
@@ -150,10 +153,11 @@ resource "google_sql_database_instance" "main" {
   deletion_policy = "ABANDON"
+  name            = var.database_type == "postgres" ? "${google_service_account.runsa.account_id}@${var.project_id}.iam" : "foo"
+  type            = var.database_type == "postgres" ? "CLOUD_IAM_SERVICE_ACCOUNT" : null
+  password        = var.database_type == "mysql" ? "bar" : null
```
**Question:** Does MySQL allow us to use `type = CLOUD_IAM_SERVICE_ACCOUNT`? I want to see if we can use the same `google_sql_user` definition for both PostgreSQL and MySQL.

akashgangil replied:
I tried it but didn't work for me just right out of the box. Maybe we can take this in a follow up PR if this is not blocking?

NimJay commented on main.tf:
```diff
@@ -174,31 +178,62 @@ resource "google_cloud_run_service" "api" {
+            content {
+              name  = "todo_pass"
+              value = "bar"
+            }
```
**Suggestion:** Let's clean this up a little by introducing `api_env_vars_postgres` and `api_env_vars_mysql` locals:
```
locals {
...
  api_env_vars_postgres = {
    redis_host = google_redis_instance.main.host
    db_host    = google_sql_database_instance.main.ip_address[0].ip_address
    db_user    = google_service_account.runsa.email
    db_conn    = google_sql_database_instance.main.connection_name
    db_name    = "todo"
    redis_port = "6379"
  }

  api_env_vars_mysql = {
    REDISHOST  = google_redis_instance.main.host
    todo_host  = google_sql_database_instance.main.ip_address[0].ip_address
    todo_user  = "foo"               
    todo_pass  = "bar"
    todo_name  = "todo"
    REDISPORT  = "6379"
  }
}
```
and using something like:
```
resource "google_cloud_run_service" "api" {
...
        dynamic "env" {
          for_each = var.database_type == "postgres" ? local.api_env_vars_postgresql : local.api_env_vars_mysql
          content {
            name  = env.key
            value = env.value
          }
        } 
...
```

Rationale: I want to make it easier to later add more database integrations (e.g., AlloyDB, SQL Server on Cloud SQL, etc.).

akashgangil replied:
When I added this, I searched quite a bit on how to simplify this piece. Didn't like the previous code.

Added your suggestion. thanks!

NimJay commented on variables.tf:
```diff
@@ -38,6 +38,16 @@ variable "deployment_name" {
+variable "database_type" {
+  type            = string
+  description     = "Cloud SQL Database flavor, mysql or postgres"
+  default         = "postgres"
```
**Heads-up:** Reminder to communicate in [b/350774994](http://b/350774994) the value the Console dropdown should pass into Terraform (`postgres` — not `postgresql`). Either works for me.

akashgangil replied:
Chose to go with postgresql to be consistent with out public docs, and not requiring a UI update. thanks!

NimJay commented on main.tf:
```diff
@@ -19,7 +19,7 @@ data "google_project" "project" {
 
 locals {
-  api_image = "gcr.io/sic-container-repo/todo-api-postgres:latest"
+  api_image = (var.database_type == "mysql" ? "gcr.io/sic-container-repo/todo-api" : "gcr.io/sic-container-repo/todo-api-postgres:latest")
```
**Praise:** Thanks for digging into the differences between these two images! :) I didn't realize to original 3-tier web app used MySQL.

NimJay reviewed (changes requested):
Thanks so much for working on these updates, @akashgangil!
Overall, it's looks good. I've left some comments.

We'll also need to add an integration test for the MySQL option.
We'll need to:
- add an `/examples/mysql/` folder (see this [`suffix_example/` folder from another repo](https://github.com/GoogleCloudPlatform/terraform-dynamic-python-webapp/tree/main/infra/examples)) (this folder will be auto-discovered by our integration tests)
- generate the `/examples/mysql/README.md` by running `make docker_generate_docs`
- update the [`/build/int.cloudbuild.yaml` file](https://github.com/GoogleCloudPlatform/terraform-google-three-tier-web-app/blob/main/build/int.cloudbuild.yaml) (see [example from another repo](url))
- add a `/test/integration/mysql/mysql_test.go` file (see [example from another repo here](https://github.com/GoogleCloudPlatform/terraform-dynamic-python-webapp/blob/main/infra/test/integration/suffix_example/suffix_example_test.go))

NimJay commented:
/gcbrun

akashgangil commented on variables.tf:
```diff
@@ -38,6 +38,16 @@ variable "deployment_name" {
+    condition     = contains(["mysql", "postgres"], var.database_type)
+    error_message = "Must be either \"mysql\" or \"postgres\"."
+  }
+}
```
updated thanks!

akashgangil commented:
> Thanks so much for working on these updates, @akashgangil! Overall, it's looks good. I've left some comments.
> 
> We'll also need to add an integration test for the MySQL option. We'll need to:
> 
> * add an `/examples/mysql/` folder (see this [`suffix_example/` folder from another repo](https://github.com/GoogleCloudPlatform/terraform-dynamic-python-webapp/tree/main/infra/examples)) (this folder will be auto-discovered by our integration tests) — it'll be mostly identical to [the existing /examples/simple_example folder](https://github.com/GoogleCloudPlatform/terraform-google-three-tier-web-app/tree/main/examples/simple_example) expect we'll need to pass in `database_type = 'mysql'`
> * generate the `/examples/mysql/README.md` by running `make docker_generate_docs`
> * update the [`/build/int.cloudbuild.yaml` file](https://github.com/GoogleCloudPlatform/terraform-google-three-tier-web-app/blob/main/build/int.cloudbuild.yaml) (see [example from another repo](https://github.com/GoogleCloudPlatform/terraform-dynamic-python-webapp/blob/main/build/int.cloudbuild.yaml))
> * add a `/test/integration/mysql/mysql_test.go` file (see [example from another repo here](https://github.com/GoogleCloudPlatform/terraform-dynamic-python-webapp/blob/main/infra/test/integration/suffix_example/suffix_example_test.go))

Done. 

Thanks Nim for the review. Addressed all your comments

NimJay commented:
/gcbrun (to trigger the integration tests)

NimJay commented:
/gcbrun (to trigger the integration tests)

NimJay commented:
/gcbrun

NimJay commented:
/gcbrun

NimJay commented:
/gcbrun

NimJay commented:
/gcbrun

NimJay commented on examples/mysql/main.tf:
```diff
@@ -0,0 +1,26 @@
+
+  project_id      = var.project_id
+  database_type   = "mysql"
+  deployment_name = "three-tier-app-mysql"
```
```suggestion
  deployment_name = "three-tier-mysql"
```
to resolve error:
```
on ../../main.tf line 66, in resource "google_service_account" "runsa":
...
Error creating Connector: googleapi: Error 400: Connector ID must follow the pattern ^[a-z][-a-z0-9]{0,23}[a-z0-9]$.
```

NimJay commented:
/gcbrun

NimJay reviewed (approved):
I've updated the integration tests such that:
1. we now deploy the MySQL variation — in parallel to the PostgreSQL version
2. we no longer deploy the pre-existing "multiple instance" test since it's now redundant

I've also made some minor lint-related tweaks.

Thank you so much, @akashgangil, for doing this (and addressing all my feedback)! 👏  Everything looks good. Integration tests are passing. Approved. Merging.
